proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,223 @@
+* Proposal: [0010](0010-resource-element-type-validation.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: Joshua Batista
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`.
+All resources can be placed in two categories, raw buffers and typed buffers. 

```

Comment:
This document would be much clearer if it focused only on HLSL "typed resource element" type constraints.

Raw buffers don't have the same sort of element type, at least not in the same sense.  In fact, strictly speaking "raw buffers" are ByteAddressBuffers, which don't have an element type at all (are treated as raw bits).  You can either use the ordinary Load[2-4]/Store methods which operate on uint1-4 granularity, or use the templated versions which allow you to specify your current interpretation of bits directly for an operation.  That's still not the "element type" of the resource.

For structured buffers, these do have a structure type, also known as an element type, but again, not in the same sense as for typed resources.

I think it's more confusing to try to describe both side-by-side with the same element terminology throughout the same document.

The constraints applied to the raw/structured buffer accesses are actually more broadly applicable than a "resource element type".  For instance, the same rules apply for raytracing or mesh payload structures, ray tracing attribute structures, groupshared variable types, ConstantBuffer/TextureBuffer template types, and really any structure that can be used to load or store host memory outside the typed resource case.

Reply:
Not to mention that this does not describe "all resources" as stated.

Reply:
Ok, I'll change the spec so that it makes no mention of raw buffers.

Reply:
Is this really necessary?  Do we need to further churn the spec for this?

Maybe "All resources" could be rewritten as "All buffer resources" if there's concern about that, but I don't know if we need to do another big reworking of this document.

Reply:
Yeah, I think the raw buffers should stay in. Their element types need to be validated by `!__builtin_hlsl_is_intangible`.